### PR TITLE
e2e: Support upgrade tests with the "local" cluster provider

### PR DESCRIPTION
**Pull request by adamwg**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

Allow upgrade tests to be run on local clusters by allowing callers to provide a script that will upgrade the cluster. Since every local cluster will be different we can't provide a default script, so the default is to throw an error - the current behavior when running upgrade tests on a local cluster. This will allow for running e2e tests against, e.g., managed kubernetes providers that support upgrades via an API.

Aside from adding the upgrade script option, I also needed to clean up one spot where the `Setup` function of an upgrade test was causing the entire suite to be skipped. Moving the skip condition to a `Skip` function solved the problem.

I've also included a change that was necessary to get these tests running on DigitalOcean Kubernetes: the PV tests were creating a 1-byte volume, which DO doesn't support. I've changed them to create a 1GiB volume, which is our minimum. I think 1GiB is a reasonable default that will work for other providers and isn't onerous, but we could parameterize this instead if anyone's concerned about the change.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

/sig testing
/sig cluster-lifecycle

### CODE REVIEW

File: test/e2e/framework/statefulset_utils.go
Comment (SataQiu) [inline test/e2e/framework/statefulset_utils.go]: What is the reason for this change ?

File: test/e2e/upgrades/storage/persistent_volumes.go
Comment (SataQiu) [inline test/e2e/upgrades/storage/persistent_volumes.go]: Thanks for your contribution! @adamwg 
Can you explain the reason for this revision?
Because I can't find out where this function is being called.

**Comment by adamwg** [comment]
/retest

Fixed the golint error that the tests caught.

File: test/e2e/framework/statefulset_utils.go
Comment (adamwg) [reply test/e2e/framework/statefulset_utils.go]: As described in the PR description, many persistent volume classes (including the default ones on DigitalOcean manage Kubernetes and EKS) don't support creating a 1-byte volume like this was previously doing. Both DO and EKS have a 1GiB minimum, so that's what I've used here. As I mentioned, we could parameterize this for more flexibility but 1GiB doesn't seem like an onerous requirement.

File: test/e2e/upgrades/storage/persistent_volumes.go
Comment (adamwg) [reply test/e2e/upgrades/storage/persistent_volumes.go]: `Skip` gets called by the [upgrade test framework here](https://github.com/adamwg/kubernetes/blob/local-upgrade-tests/test/e2e/lifecycle/cluster_upgrade.go#L388), to determine whether a particular test should be skipped. Previously this test was [being skipped from the Setup function](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/upgrades/storage/persistent_volumes.go#L55) if the provider wasn't gce or gke, but that caused the entire upgrade suite to be skipped, not just this portion.

**Comment by oomichi** [comment]
Thanks for doing this, it is an interesting feature.

from the PR message

> This will allow for running e2e tests against, e.g., managed kubernetes providers that support upgrades via an API.

`local` provider is for the situation the test target Kubernetes cluster and e2e test runner exist on the same machine as `local`. If we want to run e2e test against a remote machine of the cluster, the provider is `skeleton` unless `aws`, `gke`, `gce` or so on. Is it fine to run your upgrade test script only for `local` cluster?

**Comment by adamwg** [comment]
@oomichi Thanks for the pointer - I hadn't come across the `skeleton` provider. It's not 100% clear to me from the code and docs what the difference between `local` and `skeleton` is intended to be. I've been using `local` to run against managed clusters because (a) it was the only non-provider-specific option mentioned in the [contributors e2e testing docs](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-testing/e2e-tests.md), which don't mention skeleton at all, and (b) it matches how [sonobuoy runs the conformance tests against arbitrary k8s clusters](https://github.com/heptio/kube-conformance/blob/c4eb36ad07a5d7d7e0a47bb6535f787d589b729a/Dockerfile#L34).

From what I can find in the e2e code, the only difference between `local` and `skeleton` is how the framework finds SSH keys for connecting to nodes. In the case of managed clusters, it's unlikely that the framework can SSH to nodes anyway. Perhaps this suggests there should be another provider called `managed` or something that doesn't assume SSH access to nodes. That feels like a bigger discussion/project.

I suspect either `local` or `skeleton` will work fine for my use-case (running the e2e upgrade tests against DigitalOcean managed Kubernetes ([DOKS](https://www.digitalocean.com/products/kubernetes/)) clusters), so I'd be happy to support either or both in these changes. I'd defer to someone with more context on what the desired difference between local and skeleton is :-).

File: test/e2e/framework/nodes_util.go
Comment (timothysc) [inline test/e2e/framework/nodes_util.go]: Not a fan of this idea.  
/assign @timothysc 
/hold

**Review by timothysc — CHANGES_REQUESTED** [review]
Do you have the specific list of tests you are trying to enable with this change?

**Comment by adamwg** [comment]
@timothysc My goal with this change is to run the [`Feature:Upgrade` tests](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/lifecycle/cluster_upgrade.go#L84) against arbitrary (i.e., non-GKE/GCE) clusters. These tests deploy a set of workloads to a cluster, then trigger an upgrade of the cluster and ensure that workloads behave as expected during and after the upgrade. They're very helpful for (1) ensuring that an automated upgrade process works as expected, and (2) validating upgrade paths.

File: test/e2e/framework/nodes_util.go
Comment (adamwg) [reply test/e2e/framework/nodes_util.go]: Can you elaborate? I've implemented this in a way that mirrors how upgrades are run in the e2e tests for [GCE](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/nodes_util.go#L78) and [GKE](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/nodes_util.go#L113) clusters, but I'm happy to make changes if you can suggest a better way to accomplish the goal of running upgrade tests against arbitrary clusters.

Comment (timothysc) [reply test/e2e/framework/nodes_util.go]: The current process of how upgrades are managed in testing today are cobbled in test-infra today.  

It's much saner todo multi-step process out of the core e2es.

Comment (adamwg) [reply test/e2e/framework/nodes_util.go]: Thanks for elaborating, and apologies for my delay in coming back to this (this work has fallen down my stack of in-progress things).

I definitely agree - the upgrade process in the tests feels cobbled together at the moment. I'm happy to help alleviate that as part of or as a prerequisite for this change.

That said, I'm not sure I understand what you mean by "out of the core e2es". Is the idea that the upgrade tests themselves should manage the upgrade process, leaving the core framework out of it? Or are you suggesting that upgrade-related tests move out of the e2e framework entirely?

**Comment by timothysc** [comment]
So there is nothing that prevents you from creating tooling that does your upgrade out of band from the e2es and verifies by running as a next step.  Building upgrade (which is really a multi-step process) is a hack today which is only used by the release team.  SIG-Cluster lifecycle does upgrade tests in the method outlined above and it is the recommended approach.  

closing this issue

**Pull request closed without merging**
